core/src/vmem.rs

```diff
@@ -55,7 +55,11 @@ impl Debug for VirtualMemory {
 
 impl VirtualMemory {
     /// The maximum allocation size for a `wasmi` virtual memory.
-    const MAX_ALLOCATION_SIZE: usize = u32::MAX as usize + 1;
+    #[cfg(target_pointer_width = "64")]
+    const MAX_ALLOCATION_SIZE: usize = 2 << 32;
```

Comment:
`2 << 32` is not equal to `u32::MAX as usize + 1` but double that sice, therefore 8GB.
I'd propose to use `u32::MAX as usize + 1` instead since that's more clear.

How about this:

```rust
/// The maximum allocation size for a `wasmi` virtual memory on 16-bit.
#[cfg(target_pointer_width = "16")]
const MAX_ALLOCATION_SIZE: usize = compile_error!(
    "16-bit architectures are current unsupported by wasmi"
);

/// The maximum allocation size for a `wasmi` virtual memory on 32-bit.
#[cfg(target_pointer_width = "32")]
const MAX_ALLOCATION_SIZE: usize = i32::MAX as u32 as usize; // 2GB

/// The maximum allocation size for a `wasmi` virtual memory on 64-bit.
#[cfg(target_pointer_width = "64")]
const MAX_ALLOCATION_SIZE: usize = u32::MAX as usize + 1; // 4GB
```

With the current code and spec Wasm on 16-bit platforms is simply impossible to support properly.

Reply:
It can actually be even `i32::MAX as usize + 1`. :)

I updated the code snippet above.